File: homework3/Satyam_Mishra/Q8_AlternatingPath.java
// DS/Algo used - BFS

// Time Complexity = O(V+E)
// Space Complexity = O(V+E)

package homework3.Satyam_Mishra;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

class Edge {
    String to;
    String color;

    public Edge(String to, String color) {
        this.to = to;
        this.color = color;
    }
}

class State {
    String node;
    String lastColor;
    int distance;

    public State(String node, String lastColor, int distance) {
        this.node = node;
        this.lastColor = lastColor;
        this.distance = distance;
    }
}

public class Q8_AlternatingPath {
    private static Map<String, List<Edge>> buildadjMap(String[][] edges) {
        Map<String, List<Edge>> adj = new HashMap<>();
        for (String[] edge : edges) {
            String u = edge[0];
            String v = edge[1];
            String color = edge[2];

            adj.putIfAbsent(u, new ArrayList<>());
            adj.putIfAbsent(v, new ArrayList<>());

            adj.get(u).add(new Edge(v, color));
        }

        return adj;
    }

    public static int bfs(String[][] edges, String origin, String destination, Map<String, List<Edge>> adjMap) {
        Queue<State> q = new ArrayDeque<>();
        
        Set<String> visited = new HashSet<>();

        q.add(new State(origin, "none", 0));
        visited.add(origin + "_none");

        while(!q.isEmpty()) {
            State front = q.poll();

            if (front.node.equals(destination)) {
                return front.distance;
            }

            if (!adjMap.containsKey(front.node)) continue;

            for (Edge nei : adjMap.get(front.node)) {
                if (front.lastColor.equals("none") || !nei.color.equals(front.lastColor)) {
                    
                    String stateKey = nei.to + "_" + nei.color;
                    
                    if (!visited.contains(stateKey)) {
                        visited.add(stateKey);
                        q.add(new State(nei.to, nei.color, front.distance + 1));
                    }
                }
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        String[][] edges = {
            {"A", "B", "blue"},
            {"A", "C", "red"},
            {"B", "D", "blue"},
            {"B", "E", "blue"},
            {"C", "B", "red"},
            {"D", "C", "blue"},
            {"A", "D", "red"},
            {"D", "E", "red"},
            {"E", "C", "red"}
        };

        Map<String, List<Edge>> adjMap = buildadjMap(edges);

        System.out.println("Shortest alternating path A -> E: " + bfs(edges, "A", "E", adjMap)); 

        System.out.println("Shortest alternating path E -> D: " + bfs(edges, "E", "D", adjMap));
    }
}

// time taken 37m

File: homework3/Mudit_Sharma/q_8_AlternatingPath.cpp
//Question 8: AlternatingPath
//Given an origin and a destination in a directed graph in which edges can be blue or red, determine the 
//length of the shortest path from the origin to the destination in which the edges traversed alternate in 
//color. Return -1 if no such path exists.

// Examples: 
// [(A, B, "blue"), (A, C, "red"), (B, D, "blue"), (B, E, "blue"), (C, B, "red"), (D, C, "blue"), (A, D, "red"), (D, E, "red"), (E, C, 
// "red")] 
// Input: origin = A, destination = E 
// Output: 4 (path: A→D (red), D→C (blue), C→B (red), B→E (blue)) 
// Input: origin = E, destination = D 
// Output: -1 (only path is: E→C (red), C→B (red), B→D (blue))

//Used the concept of Disjoint set

//time complexity  :O(n*m)
//space complexity :O(n) in queue

#include <bits/stdc++.h>

// color: 1 = blue, 2 = red
int shortestAlternatingPath(
    std::vector<std::pair<std::pair<char,char>, std::string>>& edges,
    char origin,
    char destination
) {
    // build adjacency list
    std::unordered_map<char, std::vector<std::pair<char,int>>> adj;

    for (size_t i = 0; i < edges.size(); i++) {
        char u = edges[i].first.first;
        char v = edges[i].first.second;
        std::string color = edges[i].second;

        int c = (color == "blue") ? 1 : 2;
        adj[u].push_back(std::make_pair(v, c));
    }

    // visited[node][color]
    std::unordered_map<char, std::vector<bool>> visited;
    for (auto it = adj.begin(); it != adj.end(); ++it) {
        visited[it->first] = std::vector<bool>(3, false);
    }

    // queue stores: (node, lastColor, distance)
    std::queue<std::tuple<char,int,int>> q;
    q.push(std::make_tuple(origin, 0, 0));

    while (!q.empty()) {
        std::tuple<char,int,int> curr = q.front();
        q.pop();

        char node = std::get<0>(curr);
        int lastColor = std::get<1>(curr);
        int dist = std::get<2>(curr);

        if (node == destination) {
            return dist;
        }

        for (size_t i = 0; i < adj[node].size(); i++) {
            char next = adj[node][i].first;
            int edgeColor = adj[node][i].second;

            if (edgeColor != lastColor && !visited[next][edgeColor]) {
                visited[next][edgeColor] = true;
                q.push(std::make_tuple(next, edgeColor, dist + 1));
            }
        }
    }

    return -1;
}

int main() {
    std::vector<std::pair<std::pair<char,char>, std::string>> edges = {
        {{'A','B'}, "blue"},
        {{'A','C'}, "red"},
        {{'B','D'}, "blue"},
        {{'B','E'}, "blue"},
        {{'C','B'}, "red"},
        {{'D','C'}, "blue"},
        {{'A','D'}, "red"},
        {{'D','E'}, "red"},
        {{'E','C'}, "red"}
    };

    std::cout << shortestAlternatingPath(edges, 'A', 'E') << std::endl; // 4
    std::cout << shortestAlternatingPath(edges, 'E', 'D') << std::endl; // -1

    return 0;
}
